File: SistemaBancario/banco.cpp
#include "banco.h"

#include <iostream>

using namespace std;

Banco::Banco() 
{
    
    this -> numeroAtualdeContas = 4;
    this-> pContas = new Conta[this -> numeroAtualdeContas];
    
    this -> pContas[0] = {1234, 1, "Joao", "Corrente", 300};
    this -> pContas[1] = {4567, 2, "Jose", "Poupanca", 800};
    this -> pContas[2] = {7890, 3, "Maria", "Corrente", 1000};
    this -> pContas[3] = {8956, 4, "Madalena", "Poupanca", 2000};
    
    
    
    
    this->gerentes[0] = Gerente(6655,1,"Bernardo");
    this->gerentes[1] = {0405,2,"Moyses"};
    
    
}

Banco::~Banco()
{
}

Conta *Banco::buscaConta(int numero)
{
    for (int i = 0; i < this -> numeroAtualdeContas; i++)
    {
        if (numero == this->pContas[i].numero)
        {
            return &this->pContas[i];
        }
    }

    return nullptr;
}

Gerente* Banco::buscaGerente(int numero)
{
    for (int i = 0; i < NUMGERENTES; i++)
    {
        if (numero == this->gerentes[i].numero)
        {
            return &this->gerentes[i];
        }
    }

    return nullptr;
}



void Banco::atendimentoCliente() 
{
    Conta *contaCliente;
    int numC = 0;
    int senhain;
    bool atendimento = true;
    int numcontadestino;
    Conta *contaDestino;

    cout << "Bem vindo ao sistema de atendimento ao Cliente" << endl;
    cout << "Digite o numero da sua conta: ";
    cin >> numC;

    contaCliente = this->buscaConta(numC); 

    if (contaCliente == nullptr)
    {
        cout << "Conta invalida" << endl;
    }
    else
    {
        cout << "Digite a sua senha: ";
        cin >> senhain;

        if (contaCliente->validaSenha(senhain))
        {
            cout << "Ola " << contaCliente->titular << endl;
            while (atendimento) 
            {
                int op;
                double valor;
                bool status_transf = true;
                cout << "Qual operacao deseja fazer? (1 - Saque, 2 - Deposito, 3 - Ver Saldo, 4 - Transferencia, 5 - Sair): ";
                cin >> op;
                switch (op)
                {
                case 1:
                    cout << "Digite o valor: ";
                    cin>>valor;
                    contaCliente->saque(senhain,valor);
                    break;
                case 2:
                    cout << "Digite o valor: ";
                    cin>>valor;
                    contaCliente->deposito(valor);
                    break;
                case 3:
                    cout << "Saldo: R$ "<<contaCliente->getSaldo(senhain)<<endl;
                    break;
                case 4:
                    cout<< "Digite o número da conta de destino: ";
                    cin>>numcontadestino;
                    if(numcontadestino <= this->numeroAtualdeContas){
                        contaDestino = this->buscaConta(numcontadestino);
                        cout<< "Digite o valor a ser transferido: ";
                        cin>>valor;
                        status_transf = contaCliente->transferencia(valor,senhain,contaDestino);
                        if(status_transf){
                            cout<<"Transferencia de R$"<<valor<<" realizada com sucesso para conta "<<numcontadestino<<"."<<endl;
                        }
                        else{
                            cout<<"Valor de Transferencia inválido "<<endl;
                        }
                        break;
                    }
                    else{
                        cout<<"O número da conta é inválido. "<<endl;
                        break;
                    }
                
                case 5:
                    atendimento = false;
                    break;
                }
            }
        }
        else
        {
            cout << "Senha invalida" << endl;
        }
    }
}
void Banco::atendimentoGerente()
{
    int numerogerente;
    int senhain;
    bool atendimento = true;
    Gerente* contaGerente;
    
    cout<<"Bem vindo Gerente!"<<endl;
    cout<<"Digite o seu número de funcionário: "<<endl;
    cin>>numerogerente;
    
    contaGerente = this->buscaGerente(numerogerente); 
    
    if (contaGerente == nullptr)
    {
        cout << "Conta invalida" << endl;
    }
    else
    {
        cout << "Digite a sua senha: ";
        cin >> senhain;

        if (contaGerente->validaSenha(senhain))
        {
            cout << "Ola " << contaGerente->nome << endl;
            while (atendimento) 
            {
                int op;
                cout << "Qual operacao deseja fazer? (1 - Criar nova Conta, 2 - Sair): ";
                cin >> op;
                switch (op)
                {
                case 1:
                    this ->criaContaNova(senhain,contaGerente);
                    break;
                case 2:
                    atendimento = false;
                    break;
                }
            }
        }
        else
        {
            cout << "Senha invalida" << endl;
        }
    }
}

void Banco::atendimento(){
    int escolha;
    bool status_atendimento = true;
    bool atendimento_geral = true;
    while(atendimento_geral){
        cout<<"Bem vindo ao sistema de atendimento do banco "<<endl;
        cout<<"O que você deseja: 1- Atendimento ao Cliente, 2- Atendimento ao Gerente, 3 - Sair"<<endl;
        cin>>escolha;
    
        switch(escolha){
            while(status_atendimento){
                case 1:
                this -> atendimentoCliente();
                break;
            case 2:
                this -> atendimentoGerente();
                break;
            case 3:
                status_atendimento = false;
                atendimento_geral = false;
                break;
            }
        }
    }
}

int Banco::getNumeroAtualContas(int senha,Gerente* contaGerente){
    if (contaGerente->validaSenha(senha)){
        return this->numeroAtualdeContas;
    }
    else{
        cout<<"Senha inválida";
        return -10000;
    }
}

void Banco::criaContaNova(int senha, Gerente* contaGerente){
    int senhaConta;
    string nometitular;
    string tipoConta;
    double saldoConta;
    int numeroContas;
    
    numeroContas = getNumeroAtualContas(senha,contaGerente);
    pcopiaContas = new Conta[(numeroContas) + 1];
    
    for(int i = 0; i<numeroContas ; i++){
        pcopiaContas[i] = pContas[i];
    }
    
    delete[] this->pContas;
    cout<<"Entre com os dados da nova conta: "<<endl;
    cout<<"Digite o titular: "<<endl;
    cin>>nometitular;
    
    
    
    pcopiaContas[numeroContas].titular = nometitular;
    
    
    cout<<"Digite o tipo de conta: "<<endl;
    cin>>tipoConta;
    
    pcopiaContas[numeroContas].tipo = tipoConta;
    
    cout<<"Digite o saldo da conta: "<<endl;
    cin>>saldoConta;
    
    pcopiaContas[numeroContas].setSaldo(saldoConta);
    
    
    cout<<"Digite a senha da conta: "<<endl;
    cin>>senhaConta;
    
    pcopiaContas[numeroContas].setSenha(senhaConta);
    
    
    pcopiaContas[numeroContas].numero = (numeroContas + 1);
        
    cout<<"Conta criada com sucesso "<<endl;
    cout<<"Exibindo dados públicos da nova conta criada: "<<endl;
    pcopiaContas[numeroContas].exibeDados();
    
    pContas = new Conta[(numeroContas) + 1];
    
    for(int i = 0; i<(numeroContas + 1) ; i++){
        pContas[i] = pcopiaContas[i];
    }
    
    delete[] this->pcopiaContas;
    
    this->numeroAtualdeContas++;
    cout<<"Número atual de contas: "<<this->numeroAtualdeContas<<endl;
}

File: Python/Aula 10/POO/banco.py
from contas import Conta, ContaPoupanca


class Banco:
    """
    Classe Banco
    """

    def __init__(self):
        self.__contas = [
            Conta(1, senha=1234, titular="João", saldoi=500),
            ContaPoupanca(2, senha=4567, taxa=0.003, titular="José", saldoi=800),
            Conta(3, senha=7890, titular="Maria", saldoi=1000),
        ]
        self.__gerentes = [
            Conta(1, senha=6655, titular="Bernardo", saldoi=-100),
            Conta(2, senha=4455, titular="Moyses", saldoi=-100),
        ]

    def atendimento(self):
        """
        Método para atendimento geral do banco
        """
        atendimento_geral = True
        while atendimento_geral:
            print("Bem vindo ao sistema de atendimento do banco:")
            print(
                "O que você deseja: 1 - Atendimento ao Cliente, 2 - Atendimento ao Gerente, 3 - Sair"
            )
            escolha = int(input())

            match escolha:
                case 1:
                    self.atendimentoCliente()
                case 2:
                    self.atendimentoGerente()
                case 3:
                    atendimento_geral = False

    def atendimentoCliente(self):
        """
        Método para atendimento ao cliente do banco
        """
        atendimento = True
        print("Bem vindo ao sistema de atendimento ao Cliente")
        print("Digite o número da sua conta: ")
        numeroConta = int(input())
        conta_cliente = self.buscaConta(numeroConta)
        
        if numeroConta == None or conta_cliente == None:
            print("Conta inválida")
        else:
            print("Digite a sua senha: ")
            senhain = int(input())
            if conta_cliente.validaSenha(senhain):
                print("Olá ", conta_cliente.titular)
                while atendimento:
                    print(
                        "Qual operação deseja fazer? (1 - Saque, 2 - Deposito, 3 - Ver Saldo, 4 - Trocar senha, 5 - Sair)"
                    )
                    operacao = int(input())
                    match operacao:
                        case 1:
                            print("Digite o valor: ")
                            valor = float(input())
                            conta_cliente.saque(senhain, valor)
                        case 2:
                            print("Digite o valor: ")
                            valor = float(input())
                            conta_cliente.deposito(valor)
                        case 3:
                            conta_cliente.exibeDados(senhain)
                        case 4:
                            print("Digite a nova senha: ")
                            senhanova = int(input())
                            conta_cliente.setSenha(senhanova)
                        case 5:
                            atendimento = False
            else:
                print("Senha inválida")

    def atendimentoGerente(self):
        """
        Método para atendimento do gerente
        """
        atendimento = True
        print("Bem vindo Gerente!")
        print("Digite o seu número de funcionário: ")
        numeroGerente = int(input())

        if numeroGerente == None:
            print("Conta inválida")
        else:
            print("Digite a sua senha: ")
            senhaGerente = int(input())
            if self.__gerentes[numeroGerente - 1].validaSenha(senhaGerente):
                print("Olá ", self.__gerentes[numeroGerente - 1].titular)
                while atendimento:
                    print(
                        "Qual operação deseja fazer? (1 - Criar nova Conta, 2 - Remover Conta, 3 - Sair)"
                    )
                    operacao = int(input())
                    match operacao:
                        case 1:
                            self.criaContaNova(senhaGerente)
                        case 2:
                            self.removeConta(senhaGerente)
                        case 3:
                            atendimento = False
            else:
                print("Senha inválida")
    
    def buscaConta(self,numeroconta):
        for conta in self.__contas:
            if conta.numero == numeroconta:
                return conta
        return None

    def criaContaNova(self, senha):
        numeroAtualdeContas = len(self.__contas)
        print("Entre com os dados da nova conta: ")
        print("Digite o titular: ")
        nomeTitular = input()

        print("Digite o saldo da conta: ")
        saldoConta = float(input())

        print("Digite a senha da conta: ")
        senhaConta = int(input())

        numeroConta = (self.__contas[numeroAtualdeContas - 1].numero) + 1

        self.__contas.append(Conta(numeroConta, nomeTitular, senhaConta, saldoConta))
        print("Conta criada com sucesso! ")
        print("Dados da nova conta: ")
        self.__contas[numeroAtualdeContas].exibeDados(senhaConta)
        print('')

    def removeConta(self, senha):
        print("Entre com o número da conta a ser deletada: ")
        num_contaremovida = int(input())
        index = 0
        for k in self.__contas:
            if k.numero == num_contaremovida:
                break
            index += 1

        self.__contas.pop(index)
        print("Conta", num_contaremovida, "removida com sucesso")